File: Source/Game/Items/Relics/RelicTypes/FangOfShadows.cs
using System;
using WCSharp.Api;
using static WCSharp.Api.Common;

public class FangOfShadows : Relic
{
    public const int RelicItemID = Constants.ITEM_FANG_OF_SHADOWS;
    public new const int RelicAbilityID = Constants.ABILITY_SUMMON_SHADOW_KITTY;
    private const int TeleportAbilityID = Constants.ABILITY_APPEAR_AT_SHADOWKITTY;
    private new static string IconPath = "ReplaceableTextures\\CommandButtons\\BTNRingVioletSpider.blp";
    private trigger SummonTrigger;
    private trigger TeleTrigger;
    private timer KillTimer;
    private unit Owner;
    private bool Active = false;

    private const int RelicCost = 650;
    private static float SAFEZONE_REDUCTION = 0.25f; // 25%
    private static float UPGRADE_SAFEZONE_REDUCTION = 0.50f; // 50%
    private static float UPGRADE_COOLDOWN_REDUCTION = 30.0f;
    private static float SHADOW_KITTY_SUMMON_DURATION = 75.0f;

    public FangOfShadows() : base(
        $"{Colors.COLOR_PURPLE}Fang of Shadows",
        $"Ability to summon a shadowy image for {Colors.COLOR_CYAN}{(int)SHADOW_KITTY_SUMMON_DURATION} seconds|r or until death. Teleport to the illusion at will.|r " +
        $"{Colors.COLOR_ORANGE}(Active)|r {Colors.COLOR_LIGHTBLUE}(3min) (Remaining cooldown reduced by 25% at safezones.)|r",
        RelicAbilityID,
        RelicItemID,
        RelicCost,
        IconPath
        )
    {
        Upgrades.Add(new RelicUpgrade(0, $"Overall cooldown is reduced by {UPGRADE_COOLDOWN_REDUCTION} seconds.", 15, 800));
        Upgrades.Add(new RelicUpgrade(1, $"Remaining cooldown reduced at new safezones is now 50% instead of 25%.", 20, 1000));
    }

    public override void ApplyEffect(unit Unit)
    {
        Owner = Unit;
        RegisterTriggers(Unit);
        Unit.DisableAbility(RelicAbilityID, false, false);
        SetAbilityCooldown(Unit);
    }

    public override void RemoveEffect(unit Unit)
    {
        DeregisterTriggers();
        Unit.DisableAbility(RelicAbilityID, false, true);
    }

    private void RegisterTriggers(unit Unit)
    {
        SummonTrigger = trigger.Create();
        SummonTrigger.RegisterUnitEvent(Unit, unitevent.SpellCast);
        SummonTrigger.AddCondition(Condition(() => @event.SpellAbilityId == RelicAbilityID));
        SummonTrigger.AddAction(ErrorHandler.Wrap(SummonShadowKitty));

        TeleTrigger = trigger.Create();
        KillTimer = timer.Create();
    }

    private void DeregisterTriggers()
    {
        GC.RemoveTrigger(ref SummonTrigger);
        GC.RemoveTrigger(ref TeleTrigger);
        GC.RemoveTimer(ref KillTimer);
    }

    private void SummonShadowKitty()
    {
        try
        {
            Kitty summoner = Globals.ALL_KITTIES[@event.Unit.Owner];

            // Prevent summoning if holding the orb
            if (TeamDeathless.CurrentHolder == summoner)
            {
                summoner.Player.DisplayTimedTextTo(3.0f, $"{Colors.COLOR_RED}Cannot summon shadow kitty while holding the orb!{Colors.COLOR_RESET}");
                return;
            }

            ShadowKitty shadowKitty = ShadowKitty.ALL_SHADOWKITTIES[@event.Unit.Owner];

            // Summon and configure Shadow Kitty
            shadowKitty.SummonShadowKitty();
            RegisterTeleportAbility(shadowKitty.Unit);
            shadowKitty.Unit.ApplyTimedLife(FourCC("BTLF"), SHADOW_KITTY_SUMMON_DURATION);

            // Set kill timer
            KillTimer.Start(SHADOW_KITTY_SUMMON_DURATION, false, ErrorHandler.Wrap(shadowKitty.KillShadowKitty));

            // Apply relic cooldowns with a slight delay
            Utility.SimpleTimer(0.1f, () => RelicUtil.SetRelicCooldowns(Owner, RelicItemID, RelicAbilityID));
        }
        catch (Exception e)
        {
            Logger.Warning($"Error in SummonShadowKitty: {e}");
        }
    }

    private void TeleportToShadowKitty()
    {
        try
        {
            var sk = ShadowKitty.ALL_SHADOWKITTIES[@event.Unit.Owner];
            sk.TeleportToShadowKitty();
            Utility.DropAllItems(@event.Unit);
            Utility.SimpleTimer(0.09f, sk.KillShadowKitty);
            KillTimer.Pause();
        }
        catch (System.Exception e)
        {
            Logger.Warning($"Error in FangOfShadows.TeleportToShadowKitty: {e}");
            return;
        }
    }

    private void RegisterTeleportAbility(unit Unit)
    {
        TeleTrigger.RegisterUnitEvent(Unit, unitevent.SpellCast);
        TeleTrigger.AddCondition(Condition(() => @event.SpellAbilityId == TeleportAbilityID));
        TeleTrigger.AddAction(TeleportToShadowKitty);
    }

    /// <summary>
    /// Upgrade Level 1 Cooldown Reduction
    /// </summary>
    /// <param name="Unit"></param>
    private void SetAbilityCooldown(unit Unit)
    {
        var upgradeLevel = PlayerUpgrades.GetPlayerUpgrades(Unit.Owner).GetUpgradeLevel(GetType());
        var currentCooldown = BlzGetAbilityCooldown(RelicAbilityID, 0);
        var newCooldown = upgradeLevel >= 1
            ? currentCooldown - UPGRADE_COOLDOWN_REDUCTION
            : currentCooldown;

        //var ability = Unit.GetAbility(RelicAbilityID);
        RelicUtil.SetAbilityCooldown(Unit, RelicItemID, RelicAbilityID, newCooldown);
    }

    public void ReduceCooldownAtSafezone(unit Unit)
    {
        // Have relic
        if (!Utility.UnitHasItem(Unit, RelicItemID)) return;
        int upgradeLevel = PlayerUpgrades.GetPlayerUpgrades(Unit.Owner).GetUpgradeLevel(typeof(FangOfShadows));
        Unit.GetAbility(RelicAbilityID);
        float reduction = upgradeLevel >= 2 ? UPGRADE_SAFEZONE_REDUCTION : SAFEZONE_REDUCTION;
        float remainingCooldown = Unit.GetAbilityCooldownRemaining(RelicAbilityID);
        if (remainingCooldown <= 0) return;
        float newCooldown = remainingCooldown * (1.00f - reduction);
        //Unit.SetAbilityCooldownRemaining(RelicAbilityID, newCooldown);
        RelicUtil.SetRelicCooldowns(Unit, RelicItemID, RelicAbilityID, newCooldown);
    }
}

File: src/Game/Items/Relics/RelicTypes/FangOfShadows.ts
export class FangOfShadows extends Relic {
    public RelicItemID: number = Constants.ITEM_FANG_OF_SHADOWS
    public RelicAbilityID: number = Constants.ABILITY_SUMMON_SHADOW_KITTY
    private TeleportAbilityID: number = Constants.ABILITY_APPEAR_AT_SHADOWKITTY
    private static IconPath: string = 'ReplaceableTextures\\CommandButtons\\BTNRingVioletSpider.blp'
    private SummonTrigger: Trigger
    private TeleTrigger: Trigger
    private KillTimer: Timer
    private Owner: Unit
    private Active: boolean = false

    private RelicCost: number = 650
    private static SAFEZONE_REDUCTION: number = 0.25 // 25%
    private static UPGRADE_SAFEZONE_REDUCTION: number = 0.5 // 50%
    private static UPGRADE_COOLDOWN_REDUCTION: number = 30.0
    private static SHADOW_KITTY_SUMMON_DURATION: number = 75.0

    public constructor() {
        super(
            '{Colors.COLOR_PURPLE}Fang of Shadows',
            'Ability to summon a shadowy image for {Colors.COLOR_CYAN}{(int)SHADOW_KITTY_SUMMON_DURATION} seconds|r or until death. Teleport to the illusion at will.|r ' +
                '{Colors.COLOR_ORANGE}(Active)|r {Colors.COLOR_LIGHTBLUE}(3min) (Remaining cooldown reduced by 25% at safezones.)|r',
            RelicAbilityID,
            RelicItemID,
            RelicCost,
            IconPath
        )

        Upgrades.push(
            new RelicUpgrade(0, 'Overall cooldown is reduced by {UPGRADE_COOLDOWN_REDUCTION} seconds.', 15, 800)
        )
        Upgrades.push(
            new RelicUpgrade(1, 'Remaining cooldown reduced at new safezones is now 50% instead of 25%.', 20, 1000)
        )
    }

    public override ApplyEffect(Unit: Unit) {
        Owner = Unit
        this.RegisterTriggers(Unit)
        Unit.disableAbility(RelicAbilityID, false, false)
        SetAbilityCooldown(Unit)
    }

    public override RemoveEffect(Unit: Unit) {
        Dethis.RegisterTriggers()
        Unit.disableAbility(RelicAbilityID, false, true)
    }

    private RegisterTriggers(Unit: Unit) {
        let SummonTrigger = Trigger.create()!
        TriggerRegisterUnitEvent(SummonTrigger, Unit, unitevent.SpellCast)
        SummonTrigger.addCondition(Condition(() => GetSpellAbilityId() == RelicAbilityID))
        SummonTrigger.addAction(ErrorHandler.Wrap(SummonShadowKitty))

        let TeleTrigger = Trigger.create()!
        KillTimer = Timer.create()
    }

    private DeregisterTriggers() {
        GC.RemoveTrigger(SummonTrigger) // TODO; Cleanup:         GC.RemoveTrigger(ref SummonTrigger);
        GC.RemoveTrigger(TeleTrigger) // TODO; Cleanup:         GC.RemoveTrigger(ref TeleTrigger);
        GC.RemoveTimer(KillTimer) // TODO; Cleanup:         GC.RemoveTimer(ref KillTimer);
    }

    private SummonShadowKitty() {
        try {
            let summoner: Kitty = Globals.ALL_KITTIES.get(getTriggerUnit()!.owner)

            // Prevent summoning if holding the orb
            if (TeamDeathless.CurrentHolder == summoner) {
                summoner.Player.DisplayTimedTextTo(
                    3.0,
                    '{Colors.COLOR_RED}summon: shadow: kitty: while: holding: the: orb: Cannot!{Colors.COLOR_RESET}'
                )
                return
            }

            let shadowKitty: ShadowKitty = ShadowKitty.ALL_SHADOWKITTIES[getTriggerUnit().owner]

            // Summon and configure Shadow Kitty
            shadowKitty.SummonShadowKitty()
            RegisterTeleportAbility(shadowKitty.Unit)
            shadowKitty.Unit.ApplyTimedLife(FourCC('BTLF'), SHADOW_KITTY_SUMMON_DURATION)

            // Set kill timer
            KillTimer.start(SHADOW_KITTY_SUMMON_DURATION, false, ErrorHandler.Wrap(shadowKitty.KillShadowKitty))

            // Apply relic cooldowns with a slight delay
            Utility.SimpleTimer(0.1, () => RelicUtil.SetRelicCooldowns(Owner, RelicItemID, RelicAbilityID))
        } catch (e: any) {
            Logger.Warning('Error in SummonShadowKitty: {e}')
        }
    }

    private TeleportToShadowKitty() {
        try {
            let sk = ShadowKitty.ALL_SHADOWKITTIES[getTriggerUnit().owner]
            sk.TeleportToShadowKitty()
            Utility.DropAllItems(getTriggerUnit())
            Utility.SimpleTimer(0.09, sk.KillShadowKitty)
            KillTimer.pause()
        } catch (e: any) {
            Logger.Warning('Error in FangOfShadows.TeleportToShadowKitty: {e}')
            return
        }
    }

    private RegisterTeleportAbility(Unit: Unit) {
        TriggerRegisterUnitEvent(TeleTrigger, Unit, unitevent.SpellCast)
        TeleTrigger.addCondition(Condition(() => GetSpellAbilityId() == TeleportAbilityID))
        TeleTrigger.addAction(TeleportToShadowKitty)
    }

    /// <summary>
    /// Upgrade Level 1 Cooldown Reduction
    /// </summary>
    /// <param name="Unit"></param>
    private SetAbilityCooldown(Unit: Unit) {
        let upgradeLevel = PlayerUpgrades.GetPlayerUpgrades(Unit.owner).GetUpgradeLevel(GetType())
        let currentCooldown = BlzGetAbilityCooldown(RelicAbilityID, 0)
        let newCooldown = upgradeLevel >= 1 ? currentCooldown - UPGRADE_COOLDOWN_REDUCTION : currentCooldown

        //let ability = Unit.GetAbility(RelicAbilityID);
        RelicUtil.SetAbilityCooldown(Unit, RelicItemID, RelicAbilityID, newCooldown)
    }

    public ReduceCooldownAtSafezone(Unit: Unit) {
        // Have relic
        if (!Utility.UnitHasItem(Unit, RelicItemID)) return
        let upgradeLevel: number = PlayerUpgrades.GetPlayerUpgrades(Unit.owner).GetUpgradeLevel(typeof FangOfShadows)
        Unit.GetAbility(RelicAbilityID)
        let reduction: number = upgradeLevel >= 2 ? UPGRADE_SAFEZONE_REDUCTION : SAFEZONE_REDUCTION
        let remainingCooldown: number = Unit.GetAbilityCooldownRemaining(RelicAbilityID)
        if (remainingCooldown <= 0) return
        let newCooldown: number = remainingCooldown * (1.0 - reduction)
        //Unit.SetAbilityCooldownRemaining(RelicAbilityID, newCooldown);
        RelicUtil.SetRelicCooldowns(Unit, RelicItemID, RelicAbilityID, newCooldown)
    }
}
